### PR TITLE
Add package magnum

### PR DESCRIPTION
**Pull request by pthom**
* I've followed [this guide](https://docs.hunter.sh/en/latest/creating-new/create/cmake.html)
  step by step carefully. **[Yes]**

* I've tested this package remotely and have excluded all broken builds.
  Here is the links to the Travis/AppVeyor with status "All passed":

  * https://ci.appveyor.com/project/pthom/hunter/builds/22136182
  * https://travis-ci.org/pthom/hunter/builds/489024275

* I have submitted CI configs to https://github.com/ingenue/hunter targeting `pkg.template` branch,
  see this merged pull request https://github.com/ingenue/hunter/pull/333

* My change will resolve this "New package" request: https://github.com/ruslo/hunter/issues/1635

Notes: 
I did not have time to wait for the completion of some previous PR merge:
* The [ingenue PR](https://github.com/ingenue/hunter/pull/333) is still waiting to be merged
* This PR is based on a [previous upgrade PR for corrade](https://github.com/ruslo/hunter/pull/1730) 
* CI is still underway, but previous tests have shown it should be ok (let's wait it to complete anyway :-)
* 
---

### CODE REVIEW

File: examples/magnum/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    cmake_policy(SET CMP0028 NEW)
+endif()
+# Enable MACOSX_RPATH by default
+if(POLICY CMP0042)
+    cmake_policy(SET CMP0042 NEW)
+endif()
+# Quoted variables should not be dereferenced
+if(POLICY CMP0054)
+    cmake_policy(SET CMP0054 NEW)
+endif()
```
Comment (mosra) [inline examples/magnum/CMakeLists.txt]: Magnum (and Corrade) now depends on CMake 3.1, which includes all the policies above, so none of the `cmake_policy()` calls are needed anymore (and `cmake_minimum_required()` should be set to 3.1 now -- the same for Corrade, I think). Current version, for reference: https://github.com/mosra/magnum-examples/blob/master/src/primitives/CMakeLists.txt

File: examples/magnum/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+project(MagnumPrimitivesExample)
+
+# Add module path in case this is project root
+if(PROJECT_SOURCE_DIR STREQUAL CMAKE_SOURCE_DIR)
+    set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${PROJECT_SOURCE_DIR}/../../modules/")
+endif()
```
Comment (mosra) [inline examples/magnum/CMakeLists.txt]: This is referencing a path that was only present in the magnum repositories, not sure if it makes any sense here.

Comment (pthom) [reply examples/magnum/CMakeLists.txt]: @mosra: Done!
Updated tests links:
https://ci.appveyor.com/project/pthom/hunter/builds/22140210
https://travis-ci.org/pthom/hunter/builds/489086544

**Comment by pthom** [comment]
CI results after last modifs 
https://ci.appveyor.com/project/pthom/hunter/builds/22144890
https://travis-ci.org/pthom/hunter/builds/489086544

**Comment by pthom** [comment]
CI Test after merge corrade last change (v2018.10 synonym):
https://ci.appveyor.com/project/pthom/hunter/builds/22189012
https://travis-ci.org/pthom/hunter/builds/489925181

File: docs/packages/pkg/magnum.rst [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    magnum
+
+.. index::
+  single: graphics ; magnum
+
+.. _pkg.magnum:
+
+magnum
+===
```
Comment (ruslo) [inline docs/packages/pkg/magnum.rst]: https://docs.hunter.sh/en/latest/creating-new/create/cmake.html#common-mistake

File: docs/packages/pkg/magnum.rst [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+-  `Main Site <https://magnum.graphics/>`__
+-  `Hunterized <https://github.com/hunter-packages/magnum>`__
+-  `Example <https://github.com/ruslo/hunter/blob/master/examples/magnum/CMakeLists.txt>`__
+-  Developed by `Vladimír Vondruš<https://github.com/mosra>`
+-  Added by `Pascal Thomet <https://github.com/pthom>`__ (`pr-1731 <https://github.com/ruslo/hunter/pull/1731>`__)
+
+usage
+-----
+
+.. code-block:: cmake
```
Comment (ruslo) [inline docs/packages/pkg/magnum.rst]: Please use code from example, see `DOCUMENTATION_START`/`DOCUMENTATION_END`:
- https://raw.githubusercontent.com/ruslo/hunter/master/docs/packages/pkg/foo.rst

File: examples/magnum/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,36 @@
+# Copyright (c) 2015, Ruslan Baratov
+# All rights reserved.
+
+cmake_minimum_required(VERSION 3.1)
```
Comment (ruslo) [inline examples/magnum/CMakeLists.txt]: Should be `3.2`:
- https://github.com/ruslo/hunter/blob/v0.23.99/examples/foo/CMakeLists.txt#L4

File: examples/magnum/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+add_executable(magnum-primitives PrimitivesExample.cpp)
+target_link_libraries(magnum-primitives PRIVATE
+    Magnum::Application
+    Magnum::GL
+    Magnum::Magnum
+    Magnum::MeshTools
+    Magnum::Primitives
+    Magnum::Shaders)
+
+install(TARGETS magnum-primitives DESTINATION ${MAGNUM_BINARY_INSTALL_DIR})
```
Comment (ruslo) [inline examples/magnum/CMakeLists.txt]: I think we don't need it, let's keep example minimal.

File: cmake/projects/magnum/hunter.cmake [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+# All rights reserved.
+
+# !!! DO NOT PLACE HEADER GUARDS HERE !!!
+
+include(hunter_add_version)
+include(hunter_cacheable)
+include(hunter_download)
+include(hunter_pick_scheme)
+include(hunter_cmake_args)
+
```
Comment (ruslo) [inline cmake/projects/magnum/hunter.cmake]: Please clean up, one empty line.

File: cmake/projects/magnum/hunter.cmake [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    PACKAGE_NAME
+    magnum
+    VERSION
+    2019.01
+    URL
+    "https://github.com/mosra/magnum/archive/v2019.01.tar.gz"
+    SHA1
+    af356adc9e83826d55bd7f438c29f98104f95fd0
+)
+
```
Comment (ruslo) [inline cmake/projects/magnum/hunter.cmake]: One empty line.

File: cmake/projects/magnum/hunter.cmake [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+# CMAKE_SYSTEM_NAME dependant default options below
+
+# Take all the plugins
+set(MAGNUM_CMAKE_ARGS
```
Comment (ruslo) [inline cmake/projects/magnum/hunter.cmake]: This variable will leak to the user space, it's better to use `_hunter_*` instead, e.g.:
- https://github.com/ruslo/hunter/blob/v0.23.99/cmake/projects/libjson-rpc-cpp/hunter.cmake#L35-L45

**Comment by pthom** [comment]
@ruslo : Many thanks for the thorough review, and apologies for the time it took, since this branch was not rebased on master when you reviewed it !
I will rebase / squash the PR and apply your requested modifications

**Comment by pthom** [comment]
@mosra : I have one question for you:

I am refactoring the magnum example CMakeLists file (`hunter/examples/magnum/CMakeLists.txt`), in order to make it an example for the hunter integration. 

```cmake
# DOCUMENTATION_START {
hunter_add_package(magnum)
find_package(Magnum CONFIG REQUIRED
    GL
    MeshTools
    Primitives
    Shaders
    Sdl2Application)

set_directory_properties(PROPERTIES CORRADE_USE_PEDANTIC_FLAGS ON)

add_executable(magnum-primitives PrimitivesExample.cpp)
target_link_libraries(magnum-primitives PRIVATE
    Magnum::Application
    Magnum::GL
    Magnum::Magnum
    Magnum::MeshTools
    Magnum::Primitives
    Magnum::Shaders)
# DOCUMENTATION_END }
```

The flag CORRADE_USE_PEDANTIC_FLAGS is out of context as far as hunter is concerned. 
Do you prefer to keep it in the example in order to make users familiar with it?

File: cmake/projects/magnum/hunter.cmake (hunk 7 above)
Comment (pthom) [reply cmake/projects/magnum/hunter.cmake]: done

File: cmake/projects/magnum/hunter.cmake (hunk 8 above)
Comment (pthom) [reply cmake/projects/magnum/hunter.cmake]: done

File: docs/packages/pkg/magnum.rst (hunk 3 above)
Comment (pthom) [reply docs/packages/pkg/magnum.rst]: Done

**Comment by mosra** [comment]
@pthom It's not critically needed and it's more for setting up useful compiler environment. I'm mentioning it in the documentation and tutorials as a "good practice", but here it's different and the example should be as terse as possible. So remove it :)

File: docs/packages/pkg/magnum.rst [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+magnum
+======
+
+magnum is a lightweight and modular C++11/C++14 graphics middleware for games and data visualization
+
+-  `Official <https://github.com/mosra/magnum>`__
+-  `Main Site <https://magnum.graphics/>`__
+-  `Hunterized <https://github.com/hunter-packages/magnum>`__
+-  `Example <https://github.com/ruslo/hunter/blob/master/examples/magnum/CMakeLists.txt>`__
+-  Developed by `Vladimír Vondruš<https://github.com/mosra>`
```
Comment (mosra) [inline docs/packages/pkg/magnum.rst]:
```suggestion
-  Developed by `Vladimír Vondruš <https://github.com/mosra>`__
```

... so it's properly turned into a link as well

**Comment by pthom** [comment]
CI Tests after review & squash:
https://ci.appveyor.com/project/pthom/hunter/builds/22257879
https://travis-ci.org/pthom/hunter/builds/491252611

**Comment by ruslo** [comment]
Testing:
- https://ci.appveyor.com/project/ingenue/hunter/builds/22258516
- https://travis-ci.org/ingenue/hunter/builds/491266994?utm_source=github_status&utm_medium=notification

**Comment by ruslo** [comment]
Uploading:
- https://ci.appveyor.com/project/ingenue/hunter/builds/22259232
- https://travis-ci.org/ingenue/hunter/builds/491283866?utm_source=github_status&utm_medium=notification

**Comment by ruslo** [comment]
Released: https://github.com/ruslo/hunter/releases/tag/v0.23.101

**Comment by mosra** [comment]
Thanks a lot, both of you! :+1:

**Pull request merged**